compiler/rustc_resolve/src/lib.rs

```diff
@@ -1138,8 +1140,13 @@ impl ResolverAstLowering for Resolver<'_> {
         self.next_node_id()
     }
 
-    fn trait_map(&self) -> &NodeMap<Vec<TraitCandidate>> {
-        &self.trait_map
+    fn take_trait_map(&mut self) -> NodeMap<Vec<TraitCandidate>> {
+        #[cfg(debug_assertions)]
+        {
+            debug_assert!(!self.took_trait_map);
+            self.took_trait_map = true;
+        }
+        std::mem::take(&mut self.trait_map)
```

Comment:
On further though - how much work would it be to make this an `Option`, and remove `took_trait_map`?